### PR TITLE
Doc code revision rebased

### PR DESCRIPTION
**Pull request by mmatera**
This PR collects all the changes that I found are needed to make #984 work, in the sense of generating all the doctests, producing the pdf documentation, and playing well with Mathics-Django.

### CODE REVIEW

File: mathics/doc/common_doc.py [hunk 1]
```diff
@@ -502,8 +502,18 @@ def __init__(self, part, title, doc=None):
...
 
     def __str__(self) -> str:
-        sections = "\n".join(section.title for section in self.sections)
-        return f"= {self.part.title}: {self.title} =\n\n{sections}"
+        """
+        A DocChapter is represented as the index of its sections
+        and subsections.
+        """
+        sections_descr = ""
+        for section in self.all_sections:
```
Comment (mmatera) [inline mathics/doc/common_doc.py:510]: This change is optional. I just use it for testing propose.

File: mathics/doc/common_doc.py [hunk 2]
```diff
@@ -952,24 +988,327 @@ class Documentation:
...
-        self.doc_class = DocumentationEntry
-        self.doc_dir = settings.DOC_DIR
-        self.guide_section_class = DocGuideSection
-        self.guide_sections: List[DocGuideSection] = []
-        self.part_class = DocPart
-        self.parts: list[DocPart] = []
+        # This is a way to load the default classes
+        # without defining these attributes as class
+        # attributes.
+        self._set_classes()
```
Comment (mmatera) [inline mathics/doc/common_doc.py:994]: Using this mechanism seems better that passing the classes as initialization parameters. Then, each class can have its own strategy to set the attributes.

File: mathics/doc/common_doc.py [hunk 3]
```diff
@@ -952,24 +988,327 @@ class Documentation:
...
-        self.subsection_class = DocSubsection
-        self.title = "Overview"
+        self.title = "Title"
+
+    def _set_classes(self):
+        """
+        Set the classes of the subelements. Must be overloaded
+        by the subclasses.
+        """
+        if not hasattr(self, "part_class"):
```
Comment (mmatera) [inline mathics/doc/common_doc.py:1005]: The if allows to set these classes as class attributes in subclasses.

File: mathics/doc/common_doc.py [hunk 4]
```diff
@@ -1211,7 +1624,7 @@ def load_documentation_sources(self):
         # Now extract Appendix information. This include License text.  This is
         # the final Part of the documentation.
 
-        for part in appendix:
+        for part in self.appendix:
```
Comment (mmatera) [inline mathics/doc/common_doc.py:1627]: In order to decompose the loading in subtasks, I need to put the appendix variable as a class variable.

File: mathics/doc/common_doc.py [hunk 5]
```diff
@@ -1480,29 +1872,39 @@ class DocumentationEntry:
...
-        self,
-        doc_str: str,
-        title: str,
-        section: [DocSection],
-        doctests_class=DocTests,
-        doctest_class=DocTest,
-        doctext_class=DocText,
-    ):
+    def __init__(self, doc_str: str, title: str, section: Optional[DocSection] = None):
+        self._set_classes()
```
Comment (mmatera) [inline mathics/doc/common_doc.py:1876]: Same idea as in Documentation: specific classes must be specified in the subclasses.

File: mathics/doc/common_doc.py [hunk 6]
```diff
@@ -805,7 +841,7 @@ def __init__(
 
         # Needs to come after self.chapter is initialized since
         # XMLDoc uses self.chapter.
-        self.xml_doc = DocumentationEntry(text, title, self)
+        self.doc = DocumentationEntry(text, title, self)
```
Comment (mmatera) [inline mathics/doc/common_doc.py:844]: @rocky, we agree that "xml" is misleading, right?

**Comment by mmatera** [comment]
This PR contains the changes (including some tweaks I elaborated in #993) to make #984 work. 
After merging this, I think that still, we still need to do some extra cleaning and reordering to make the code more readable and free of dead branches. 

Also, I see that   `mathics.doc.common` module is too large. I distinguish three different things inside it that could be put in different submodules:
* The structural class hierarchy (Documentation->DocPart->DocChapter->DocSection/DocGuide->DocSubsection->DocumentationEntry)
* The DocTest elements and their collections (DocTests and Tests).
* The mechanisms to gather the Documentation from the Python libraries.

But in any case, it is work for a subsequent step.

File: mathics/doc/common_doc.py (hunk 6 above)
Comment (rocky) [reply mathics/doc/common_doc.py:844]: Yes, it is misleading or just plain wrong. 

It is possible (I hope) in the future we will hook into Sphinx's representation of a documentation object.

**Comment by rocky** [comment]
> This PR contains the changes (including some tweaks I elaborated in #993) to make #984 work. After merging this, I think that still, we still need to do some extra cleaning and reordering to make the code more readable and free of dead branches.
> 
> Also, I see that `mathics.doc.common` module is too large. I distinguish three different things inside it that could be put in different submodules:
> 
> * The structural class hierarchy (Documentation->DocPart->DocChapter->DocSection/DocGuide->DocSubsection->DocumentationEntry)
> * The DocTest elements and their collections (DocTests and Tests).
> * The mechanisms to gather the Documentation from the Python libraries.
> 
> But in any case, it is work for a subsequent step.

Yes, please I would like to see more work on getting the problems that motivated this fixed. 

With these changes, are we able to create Guide sections in Mathics3 modules now?

**Comment by mmatera** [comment]
@rocky with this changes the Guide sections seem to work in the PDF documentation (at least for pymathics.graph) and together with  mathics-django/#201 it also produces the online documentation. 

Still, there are some small details, like that in the online documentation, after loading a module, it is needed to press F5 to see the changes in the documentation, and that in the PDF documentation, sections out of a GuideSection look a little bit different. In any case, there are things that we can fix after this main part is completed.

**Comment by mmatera** [comment]
This is an example about how does it looks the pymathics.graph documentation in the PDF file:

imagen


and in the different levels of the online docs

imagen


imagen

imagen

**Comment by mmatera** [comment]
> I'll look at everything within a day or two.


Good. I will not be able to work too much on this until de weekend.

 > The screenshot above lists "Centralities" 3 times. Here is what "mathicsdoc" currently shows. (It too lists "Centralities" multiple times).


Not sure what you mean. In the screenshot I just see a repetition in the url part. In any case , I think it will take one or two more rounds to be get this ready.




> mathics3-modules
> networks-and-graphs
>

**Comment by rocky** [comment]
Superceded by other commits.

**Pull request closed without merging**